lib/consumerGroup.js

```diff
@@ -0,0 +1,455 @@
+    },
+
+    function (syncGroupResponse, callback) {
+      self.handleSyncGroup(syncGroupResponse, callback);
+    }
+  ], function (error, startFetch) {
+    self.connecting = false;
+    if (error) {
+      return self.recovery.tryToRecoverFrom(error, 'connect');
+    }
+
+    self.ready = true;
+    self.recovery.clearError();
+
+    debug('generationId', self.generationId);
+
+    if (startFetch) {
+      self.fetch();
+    }
+    self.startHearbeats();
+    self.emit('connect');
+    self.emit('rebalanced');
+  });
+};
+
+ConsumerGroup.prototype.scheduleReconnect = function (timeout) {
+  assert(timeout);
+
+  if (this.reconnectTimer) {
+    clearTimeout(this.reconnectTimer);
+  }
+
+  var self = this;
+  this.reconnectTimer = setTimeout(function () {
+    self.reconnectTimer = null;
+    self.connect();
+  }, timeout);
+};
+
+ConsumerGroup.prototype.startHearbeats = function () {
+  assert(this.options.sessionTimeout > 0);
+  assert(this.ready, 'consumerGroup is not ready');
+
+  const heartbeatIntervalMs = this.options.heartbeatInterval || (Math.floor(this.options.sessionTimeout / 3));
+
+  debug('%s started heartbeats at every %d ms', this.client.clientId, heartbeatIntervalMs);
+  var self = this;
+  this.sendHeartbeat();
+  this.hearbeatInterval = setInterval(function () {
+    self.sendHeartbeat();
+  }, heartbeatIntervalMs);
+};
+
+ConsumerGroup.prototype.stopHeartbeats = function () {
+  this.hearbeatInterval && clearInterval(this.hearbeatInterval);
+};
+
+ConsumerGroup.prototype.leaveGroup = function (callback) {
+  debug('%s leaving group', this.client.clientId);
+  var self = this;
+  this.stopHeartbeats();
+  if (self.generationId != null) {
+    this.client.sendLeaveGroupRequest(this.options.groupId, this.memberId, function (error) {
+      self.generationId = null;
+      callback(error);
+    });
+  } else {
+    callback(null);
+  }
+};
+
+ConsumerGroup.prototype.sendHeartbeat = function () {
+  assert(this.memberId, 'invalid memberId');
+  assert(this.generationId >= 0, 'invalid generationId');
+  debug('%s ❤️  ->', this.client.clientId);
+  var self = this;
+  this.client.sendHeartbeatRequest(this.options.groupId, this.generationId, this.memberId, function (error) {
+    if (error) {
+      debug('Heartbeat error:', error);
+      self.recovery.tryToRecoverFrom(error, 'heartbeat');
+    }
+    debug('%s 💚 <-', self.client.clientId, error);
+  });
+};
+
+ConsumerGroup.prototype.fetchOffset = function (payloads, cb) {
+  this.client.sendOffsetFetchV1Request(this.options.groupId, payloads, cb);
+};
+
+ConsumerGroup.prototype.autoCommit = function (force, cb) {
+  if (arguments.length === 1) {
+    cb = force;
+    force = false;
+  }
+
+  if (this.committing && !force) return cb(null, 'Offset committing');
+
+  this.committing = true;
+  setTimeout(function () {
+    this.committing = false;
+  }.bind(this), this.options.autoCommitIntervalMs);
```

Comment:
Sometimes you use bind and other times you use ```var self = this```. Should probably pick a method to be consistent to use throughout.